patches/nginx-1.13.6-socket_cloexec.patch

```diff
@@ -0,0 +1,142 @@
++                           ls[i].type | SOCK_CLOEXEC, 0);
++
++#else
+             s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type, 0);
+ 
++#if (NGX_HAVE_FD_CLOEXEC)
++            if (ngx_cloexec(s) == -1) {
++                ngx_log_error(NGX_LOG_EMERG, log, ngx_socket_errno,
++                              ngx_cloexec_n " %V failed", &ls[i].addr_text);
++                return NGX_ERROR;
++            }
++#endif
++
++#endif
++
+             if (s == (ngx_socket_t) -1) {
+                 ngx_log_error(NGX_LOG_EMERG, log, ngx_socket_errno,
+                               ngx_socket_n " %V failed", &ls[i].addr_text);
+diff --git a/src/event/ngx_event_accept.c b/src/event/ngx_event_accept.c
+index 77563709..5827b9d0 100644
+--- a/src/event/ngx_event_accept.c
++++ b/src/event/ngx_event_accept.c
+@@ -62,7 +62,9 @@ ngx_event_accept(ngx_event_t *ev)
+ 
+ #if (NGX_HAVE_ACCEPT4)
+         if (use_accept4) {
+-            s = accept4(lc->fd, &sa.sockaddr, &socklen, SOCK_NONBLOCK);
++            s = accept4(lc->fd, &sa.sockaddr, &socklen,
++                        SOCK_NONBLOCK | SOCK_CLOEXEC);
++
+         } else {
+             s = accept(lc->fd, &sa.sockaddr, &socklen);
+         }
+@@ -202,6 +204,16 @@ ngx_event_accept(ngx_event_t *ev)
+                     ngx_close_accepted_connection(c);
+                     return;
+                 }
++
++#if (NGX_HAVE_FD_CLOEXEC)
++                if (ngx_cloexec(s) == -1) {
++                    ngx_log_error(NGX_LOG_ALERT, ev->log, ngx_socket_errno,
++                                  ngx_cloexec_n " failed");
++                    ngx_close_accepted_connection(c);
++                    return;
++                }
++#endif
++
+             }
+         }
+ 
+diff --git a/src/event/ngx_event_connect.c b/src/event/ngx_event_connect.c
+index c5bb8068..484fb0bb 100644
+--- a/src/event/ngx_event_connect.c
++++ b/src/event/ngx_event_connect.c
+@@ -38,8 +38,23 @@ ngx_event_connect_peer(ngx_peer_connection_t *pc)
+ 
+     type = (pc->type ? pc->type : SOCK_STREAM);
+ 
++#if (NGX_HAVE_SOCKET_CLOEXEC)
++    s = ngx_socket(pc->sockaddr->sa_family, type | SOCK_CLOEXEC, 0);
++
++#else
+     s = ngx_socket(pc->sockaddr->sa_family, type, 0);
+ 
++#if (NGX_HAVE_FD_CLOEXEC)
++    if (ngx_cloexec(s) == -1) {
++        ngx_log_error(NGX_LOG_ALERT, pc->log, ngx_socket_errno,
++                      ngx_cloexec_n " failed");
++        return NGX_ERROR;
++    }
++#endif
++
++#endif
++
++
+     ngx_log_debug2(NGX_LOG_DEBUG_EVENT, pc->log, 0, "%s socket %d",
+                    (type == SOCK_STREAM) ? "stream" : "dgram", s);
+ 
+diff --git a/src/os/unix/ngx_socket.h b/src/os/unix/ngx_socket.h
+index fcc51533..d1eebf47 100644
+--- a/src/os/unix/ngx_socket.h
++++ b/src/os/unix/ngx_socket.h
+@@ -38,6 +38,17 @@ int ngx_blocking(ngx_socket_t s);
+ 
+ #endif
+ 
++#if (NGX_HAVE_FD_CLOEXEC)
++
++#define ngx_cloexec(s)      fcntl(s, F_SETFD, FD_CLOEXEC)
++#define ngx_cloexec_n       "fcntl(FD_CLOEXEC)"
++
```

Comment:
Better put the `#endif` here?

Reply:
@agentzh 
If `NGX_HAVE_FD_CLOEXEC` is not supported, this patch actually does nothing.
Maybe we should rename `HAVE_SOCKET_CLOEXEC_PATCH` to `ENABLE_SOCKET_CLOEXEC` to make it clear?

Reply:
@spacewander I think that macro is only an indicator for the existence of the patch itself, instead of indicating the existence of the actual feature :)

But I agree `HAVE_SOCKET_CLOEXEC_PATCH` is a bit confusing and should be renamed to `HAVE_CLOEXEC_PATCH` instead (`ENABLE_SOCKET_CLOEXEC` is actually worse, I'm afraid, which breaks our macro naming convention for such nginx core patches).